File: src/integrations/api/proctoring.ts
/**
 * Proctoring API Client
 * Production-grade API methods for proctoring features
 */

import { apiClient } from './client';

export interface ViolationCreate {
  lab_id: number;
  violation_type: 'tab_switch' | 'fullscreen_exit' | 'window_blur' | 'copy_paste' | 'devtools';
  severity: 'low' | 'medium' | 'high';
  details?: Record<string, any>;
  description?: string;
  time_spent_seconds?: number;
  problem_id?: number;
  submission_id?: number;
  timestamp?: string;
}

export interface ViolationResponse {
  id: number;
  lab_id: number;
  user_id: number;
  session_id?: number;
  violation_type: string;
  severity: string;
  timestamp: string;
  details?: Record<string, any>;
  time_spent_seconds?: number;
}

export interface SessionResponse {
  id: number;
  lab_id: number;
  user_id: number;
  started_at: string;
  total_violations: number;
  tab_switches: number;
  fullscreen_exits: number;
  is_active: boolean;
}

export interface ViolationSummary {
  total_violations: number;
  total_sessions: number;
  active_sessions: number;
  by_type: Record<string, number>;
  by_severity: Record<string, number>;
  by_user: Record<number, number>;
}

class ProctoringAPI {
  /**
   * Create or update a proctoring session
   */
  async createSession(labId: number): Promise<SessionResponse> {
    return apiClient.request(`/proctoring/sessions`, {
      method: 'POST',
      body: JSON.stringify({ lab_id: labId }),
    });
  }

  /**
   * Record a proctoring violation
   */
  async recordViolation(violation: ViolationCreate): Promise<ViolationResponse> {
    return apiClient.request(`/proctoring/violations`, {
      method: 'POST',
      body: JSON.stringify(violation),
    });
  }

  /**
   * Get all sessions for a lab (Faculty/Admin only)
   */
  async getLabSessions(labId: number, activeOnly: boolean = false): Promise<SessionResponse[]> {
    return apiClient.request(`/proctoring/sessions/${labId}?active_only=${activeOnly}`);
  }

  /**
   * Get violations for a lab (Faculty/Admin only)
   */
  async getLabViolations(
    labId: number,
    options?: {
      userId?: number;
      violationType?: string;
      severity?: string;
      limit?: number;
    }
  ): Promise<ViolationResponse[]> {
    const params = new URLSearchParams();
    if (options?.userId) params.append('user_id', options.userId.toString());
    if (options?.violationType) params.append('violation_type', options.violationType);
    if (options?.severity) params.append('severity', options.severity);
    if (options?.limit) params.append('limit', options.limit.toString());

    return apiClient.request(`/proctoring/violations/${labId}?${params.toString()}`);
  }

  /**
   * Get violation summary for a lab (Faculty/Admin only)
   */
  async getViolationSummary(labId: number): Promise<ViolationSummary> {
    return apiClient.request(`/proctoring/summary/${labId}`);
  }

  /**
   * End a proctoring session
   */
  async endSession(sessionId: number): Promise<SessionResponse> {
    return apiClient.request(`/proctoring/sessions/${sessionId}/end`, {
      method: 'PUT',
    });
  }
}

export const proctoringAPI = new ProctoringAPI();

File: backend/app/schemas/proctoring.py
"""Proctoring Schemas"""
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
from datetime import datetime
from app.models.proctoring import ViolationType, ViolationSeverity


class ViolationCreate(BaseModel):
    lab_id: int
    violation_type: str = Field(..., description="Type of violation")
    severity: str = Field(default="low", description="Severity level")
    details: Optional[Dict[str, Any]] = None
    description: Optional[str] = None
    time_spent_seconds: Optional[int] = None
    problem_id: Optional[int] = None
    submission_id: Optional[int] = None
    timestamp: Optional[datetime] = None


class ViolationResponse(BaseModel):
    id: int
    lab_id: int
    user_id: int
    session_id: Optional[int]
    submission_id: Optional[int]
    violation_type: str
    severity: str
    timestamp: datetime
    details: Optional[Dict[str, Any]] = None
    description: Optional[str] = None
    time_spent_seconds: Optional[int] = None
    problem_id: Optional[int] = None
    is_reviewed: bool = False
    reviewed_by: Optional[int] = None
    reviewed_at: Optional[datetime] = None
    review_notes: Optional[str] = None
    
    class Config:
        from_attributes = True


class SessionCreate(BaseModel):
    lab_id: int


class SessionResponse(BaseModel):
    id: int
    lab_id: int
    user_id: int
    started_at: datetime
    ended_at: Optional[datetime] = None
    total_time_seconds: int = 0
    total_violations: int = 0
    tab_switches: int = 0
    fullscreen_exits: int = 0
    window_blurs: int = 0
    copy_paste_events: int = 0
    devtools_opens: int = 0
    is_active: bool = True
    last_activity: datetime
    violation_summary: Optional[Dict[str, Any]] = None
    
    class Config:
        from_attributes = True


class ViolationSummaryResponse(BaseModel):
    total_violations: int
    total_sessions: int
    active_sessions: int
    by_type: Dict[str, int]
    by_severity: Dict[str, int]
    by_user: Dict[int, int]

